File: pythonlib/winfox/rdp/ports.py
import asyncio
import time
from typing import Optional


def _check_port(host: str, port: int) -> bool:
    """Synchronous TCP port check (Windows-compatible)."""
    import socket

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.settimeout(1.0)
    try:
        sock.connect((host, port))
        sock.close()
        return True
    except (ConnectionRefusedError, OSError, socket.timeout):
        return False


def _port_bindable(host: str, port: int) -> bool:
    import socket

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        sock.bind((host, port))
        return True
    except OSError:
        return False
    finally:
        try:
            sock.close()
        except OSError:
            pass


async def _wait_for_port(host: str, port: int, timeout: float = 60.0) -> None:
    """Wait for a TCP port to accept connections. Uses sync socket in thread for Windows compatibility."""
    deadline = time.time() + timeout
    delay = 0.2
    while time.time() < deadline:
        is_open = await asyncio.to_thread(_check_port, host, port)
        if is_open:
            return
        await asyncio.sleep(delay)
        delay = min(delay * 1.5, 2.0)
    raise TimeoutError(f"Port {port} not ready within {timeout}s")


class _PortAllocator:
    """Process-local port allocator for RDPBrowser instances."""

    def __init__(self):
        self._lock = asyncio.Lock()
        self._reserved: set[int] = set()

    async def reserve(self, port: int) -> int:
        async with self._lock:
            if port in self._reserved:
                raise RuntimeError(f"Port already reserved in allocator: {port}")
            bindable = await asyncio.to_thread(_port_bindable, "127.0.0.1", port)
            if not bindable:
                raise RuntimeError(f"Port is not available: {port}")
            self._reserved.add(port)
            return port

    async def find_and_reserve(self, start_port: int, limit: int = 500) -> int:
        async with self._lock:
            port = start_port
            while port < start_port + limit:
                if port in self._reserved:
                    port += 1
                    continue
                bindable = await asyncio.to_thread(_port_bindable, "127.0.0.1", port)
                if bindable:
                    self._reserved.add(port)
                    return port
                port += 1
            raise RuntimeError(f"Unable to find available port near {start_port}")

    async def release(self, port: Optional[int]) -> None:
        if port is None:
            return
        async with self._lock:
            self._reserved.discard(port)


_PORT_ALLOCATOR = _PortAllocator()

File: packages/winfox/src/rdp/ports.ts
import net from "net";

function _checkPort(host: string, port: number): Promise<boolean> {
  return new Promise((resolve) => {
    const sock = new net.Socket();
    sock.setTimeout(1000);
    sock.on('connect', () => {
      sock.destroy();
      resolve(true);
    });
    sock.on('timeout', () => {
      sock.destroy();
      resolve(false);
    });
    sock.on('error', () => {
      resolve(false);
    });
    sock.connect(port, host);
  });
}

function _portBindable(host: string, port: number): Promise<boolean> {
  return new Promise((resolve) => {
    const server = net.createServer();
    server.unref();
    server.on('error', () => resolve(false));
    server.listen(port, host, () => {
      server.close(() => resolve(true));
    });
  });
}

export async function waitForPort(host: string, port: number, timeout: number = 60.0): Promise<void> {
  const deadline = Date.now() + timeout * 1000;
  let delay = 200;
  while (Date.now() < deadline) {
    const isOpen = await _checkPort(host, port);
    if (isOpen) return;
    await new Promise(r => setTimeout(r, delay));
    delay = Math.min(delay * 1.5, 2000);
  }
  throw new Error(`Port ${port} not ready within ${timeout}s`);
}

class PortAllocator {
  private reserved = new Set<number>();
  private queue: Array<() => void> = [];
  private locked = false;

  private async acquireLock() {
    if (!this.locked) {
      this.locked = true;
      return;
    }
    await new Promise<void>(resolve => this.queue.push(resolve));
  }

  private releaseLock() {
    if (this.queue.length > 0) {
      const next = this.queue.shift();
      if (next) next();
    } else {
      this.locked = false;
    }
  }

  async reserve(port: number): Promise<number> {
    await this.acquireLock();
    try {
      if (this.reserved.has(port)) throw new Error(`Port already reserved in allocator: ${port}`);
      const bindable = await _portBindable("127.0.0.1", port);
      if (!bindable) throw new Error(`Port is not available: ${port}`);
      this.reserved.add(port);
      return port;
    } finally {
      this.releaseLock();
    }
  }

  async findAndReserve(startPort: number, limit: number = 500): Promise<number> {
    await this.acquireLock();
    try {
      let port = startPort;
      while (port < startPort + limit) {
        if (this.reserved.has(port)) {
          port++;
          continue;
        }
        const bindable = await _portBindable("127.0.0.1", port);
        if (bindable) {
          this.reserved.add(port);
          return port;
        }
        port++;
      }
      throw new Error(`Unable to find available port near ${startPort}`);
    } finally {
      this.releaseLock();
    }
  }

  async release(port: number | undefined): Promise<void> {
    if (port === undefined) return;
    await this.acquireLock();
    try {
      this.reserved.delete(port);
    } finally {
      this.releaseLock();
    }
  }
}

export const PORT_ALLOCATOR = new PortAllocator();
